bat.d/41-framework

```diff
@@ -0,0 +1,536 @@
+        printf "%s" "$out"
+    else
+        if [ "$2" = "1" ]; then
+            printf "(not available)\n"
+        fi
+        rc=4
+    fi
+
+    echo_debug "bat" "batdrv.${_batdrv_plugin}.read_threshold($1): bm_thresh=$_bm_thresh; bf=$bf; out=$out; rc=$rc"
+    return $rc
+}
+
+batdrv_write_thresholds () {
+    # write both charge thresholds for a battery
+    # use pre-determined method and sysfiles from global parms
+    # $1: new start threshold 0(disabled)..100/DEF(default)
+    # $2: new stop threshold 0..100/DEF(default)
+    # $3: 0=quiet/1=output parameter errors/2=output progress and errors
+    # $4: config battery - non-empty string indicates thresholds stem from configuration
+    # global param: $_bm_thresh, $_bat_str, $_bf_stop
+    # rc: 0=ok/
+    #     1=not configured/
+    #     2=threshold(s) out of range or non-numeric/
+    #     3=minimum start stop diff violated/
+    #     4=threshold read error/
+    #     5=threshold write error
+    # prerequisite: batdrv_init(), batdrv_select_battery()
+    local new_stop=${2:-}
+    local verb=${3:-0}
+    local cfg_bat="$4"
+    local old_stop
+
+    # insert defaults
+    [ "$new_stop" = "DEF" ] && new_stop=$_bt_def_stop
+
+    # --- validate thresholds
+    local rc
+
+    # TEMPLATE: customize boundaries to vendor specs
+    # stop: check for 3 digits max, ensure min 0 / max 100
+    if ! is_uint "$new_stop" 3 || \
+       ! is_within_bounds "$new_stop" 0 100; then
+        # threshold out of range
+        echo_debug "bat" "batdrv.${_batdrv_plugin}.write_thresholds($1, $2, $3, $4).invalid_stop: bat=$_bat_str"
+        case $verb in
+            1)
+                if [ -n "$cfg_bat" ]; then
+                    echo_message "Error in configuration at STOP_CHARGE_THRESH_${cfg_bat}=\"${new_stop}\": not specified, invalid or out of range (0..100). Battery skipped."
+                fi
+                ;;
+
+            2)
+                if [ -n "$cfg_bat" ]; then
+                    printf "Error in configuration at STOP_CHARGE_THRESH_%s=\"%s\": not specified, invalid or out of range (0..100). Aborted.\n" "$cfg_bat" "$new_stop" 1>&2
+                else
+                    printf "Error: stop charge threshold (%s) for %s is not specified, invalid or out of range (0..100). Aborted.\n" "$new_stop" "$_bat_str" 1>&2
+                fi
+                ;;
+        esac
+        return 2
+    fi
+
+    # read active threshold values
+    if ! old_stop=$(batdrv_read_threshold stop 0); then
+        echo_debug "bat" "batdrv.${_batdrv_plugin}.write_thresholds($1, $2, $3, $4).read_error: bat=$_bat_str"
+        case $verb in
+            1) echo_message "Error: could not read current charge threshold(s) for $_bat_str. Battery skipped." ;;
+            2) printf "Error: could not read current charge threshold(s) for %s. Aborted.\n" "$_bat_str" 1>&2 ;;
+        esac
+        return 4
+    fi
+
+    # TEMPLATE: customize assertion to vendor specs
```

Comment:
If you have a stop threshold only, it makes no sense to determine the start vs. stop sequence, just do it like here: https://github.com/linrunner/TLP/blob/main/bat.d/15-asus#L303